src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,31 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 
 public class AsyncRequestProcessor {
     private final Executor executor;
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        if (cache.containsKey(userId)) {
+            return CompletableFuture.supplyAsync(() -> cache.get(userId), executor);
+        }
+        return CompletableFuture
+                .supplyAsync(() -> new UserData(userId, "Details for " + userId), executor)
+                .thenApply(userData -> {
+                    setCache(userId, userData);
+                    return userData;
+                });
+    }
+
+    private void setCache(String id, UserData userData) {
```

Comment:
The `setCache` method is not synchronized or thread-safe. While `ConcurrentHashMap` handles some concurrent operations safely, the combination of checking for existence and then updating (lines 17-25) can lead to race conditions. Consider ensuring that the entire check-then-act sequence is atomic to prevent potential issues with concurrent access.